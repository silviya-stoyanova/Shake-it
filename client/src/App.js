import React, { Component } from 'react';
import { BrowserRouter } from 'react-router-dom'
import { ToastContainer } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css';
import './static/css/site.css'
import './static/css/scrollbar.css'

import sessionManager from './utilities/session-util'
import observer from './utilities/observer'

import MyRoutes from './components/routes/router'
import Header from './components/common/header'
import Footer from './components/common/footer'

const defaultValue = { isLogged: false }
const { Consumer: UserInfoConsumer, Provider: UserInfoProvider } = React.createContext(defaultValue)

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isLogged: sessionManager.isUserLogged(),
      username: sessionManager.getUserInfo().username,
      role: sessionManager.getUserInfo().role
    }

    observer.subscribe('userLogin', this.onUserLogin)
    observer.subscribe('userLogout', this.onUserLogout)
  }

  onUserLogin = () => {
    this.setState(prevState => ({ // asyncronous
      isLogged: true,
      username: sessionManager.getUserInfo().username,
      role: sessionManager.getUserInfo().role
    }))
  }

  onUserLogout = () => {
    this.setState(prevState => ({
      isLogged: false,
      username: '',
      role: []
    }))
  }

  render() {
    const { isLogged, username, role } = this.state

    return (
      <UserInfoProvider value={{ isLogged, username, role }}>
        <BrowserRouter>
          <ToastContainer />
          <Header />
          <MyRoutes />
          <Footer />
        </BrowserRouter>
      </UserInfoProvider>
    )
  }

  componentDidMount() {
    sessionManager.isUserLogged()
      ? this.setState({ isLogged: true })
      : this.setState({ isLogged: false })
  }
}

export { UserInfoProvider } // used in tests
export { UserInfoConsumer }
export default App