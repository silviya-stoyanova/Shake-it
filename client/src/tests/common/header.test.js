import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { mount } from 'enzyme'
import { UserInfoProvider } from '../../App'
import Header from '../../components/common/header'

// is manually wrapping MyComponent in AppContext.Provider
// in each test the only way then ?
describe('tests for the Header component', () => {
    it('should render properly when a user is NOT logged in', () => {
        //! lesson: You cannot use a Consumer from one context with Provider of another, be it with the same props passed down
        // const { Consumer, Provider } = React.createContext(defaultValue)
        const defaultValue = { isLogged: false, username: '', role: '' }

        const wrapper = mount(
            <UserInfoProvider value={defaultValue}>
                <BrowserRouter>
                    <Header />
                </BrowserRouter>
            </UserInfoProvider>
        )
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('should render properly when a user is logged in', () => {
        //! lesson: You cannot use a Consumer from one context with Provider of another, be it with the same props being passed down
        // const { Consumer: UserInfoConsumer, Provider: UserInfoProvider } = React.createContext(defaultValue)

        const defaultValue = { isLogged: true, username: 'nqkakvo-si-ime-tam', role: 'User' }

        const wrapper = mount(
            <UserInfoProvider value={defaultValue}>
                <BrowserRouter>
                    <Header />
                </BrowserRouter>
            </UserInfoProvider>
        )

        expect(wrapper.html()).toMatchSnapshot()

        expect(wrapper.find('li.greeting').length).toEqual(1)
        expect(wrapper.find('li.greeting').text()).toEqual('welcome, nqkakvo-si-ime-tam')

        expect(wrapper.find('li.logout').length).toEqual(1)
        expect(wrapper.find('li.logout').text()).toEqual('logout')
    })

    it('should render properly when The Adnin is logged in', () => {
        const defaultValue = { isLogged: true, username: 'admin', role: 'Admin' }

        const wrapper = mount(
            <UserInfoProvider value={defaultValue}>
                <BrowserRouter>
                    <Header />
                </BrowserRouter>
            </UserInfoProvider>
        )

        expect(wrapper.html()).toMatchSnapshot()

        expect(wrapper.find('li.add-product').length).toEqual(1)
        expect(wrapper.find('li.add-product').text()).toEqual('add product')

        expect(wrapper.find('li.greeting').length).toEqual(1)
        expect(wrapper.find('li.greeting').text()).toEqual('welcome, admin')

        expect(wrapper.find('li.logout').length).toEqual(1)
        expect(wrapper.find('li.logout').text()).toEqual('logout')
    })
})