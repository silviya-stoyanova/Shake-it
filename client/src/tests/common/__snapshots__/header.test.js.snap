// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for the Header component should render properly when The Adnin is logged in 1`] = `"<header><a href=\\"/\\"><span class=\\"logo\\"></span></a><ul><li class=\\"left-align\\"><a href=\\"/\\">all products</a></li><li class=\\"left-align add-product\\"><a href=\\"/product/create\\">add product</a></li><li class=\\"right-align logout\\"><a href=\\"/user/logout\\">logout</a></li><li class=\\"right-align\\"><a class=\\"my-cart-img\\" href=\\"/cart\\"></a></li><li class=\\"right-align greeting\\"><a href=\\"/user/profile\\">welcome, admin</a></li><li class=\\"left-align\\"><a href=\\"/contacts\\">contact us</a></li><li class=\\"left-align\\"><a href=\\"/about\\">about us</a></li></ul></header>"`;

exports[`tests for the Header component should render properly when a user is NOT logged in 1`] = `"<header><a href=\\"/\\"><span class=\\"logo\\"></span></a><ul><li class=\\"left-align\\"><a href=\\"/\\">all products</a></li><li class=\\"right-align\\"><a href=\\"/user/register\\">register</a></li><li class=\\"right-align\\"><a href=\\"/user/login\\">login</a></li><li class=\\"left-align\\"><a href=\\"/contacts\\">contact us</a></li><li class=\\"left-align\\"><a href=\\"/about\\">about us</a></li></ul></header>"`;

exports[`tests for the Header component should render properly when a user is logged in 1`] = `"<header><a href=\\"/\\"><span class=\\"logo\\"></span></a><ul><li class=\\"left-align\\"><a href=\\"/\\">all products</a></li><li class=\\"right-align logout\\"><a href=\\"/user/logout\\">logout</a></li><li class=\\"right-align\\"><a class=\\"my-cart-img\\" href=\\"/cart\\"></a></li><li class=\\"right-align greeting\\"><a href=\\"/user/profile\\">welcome, nqkakvo-si-ime-tam</a></li><li class=\\"left-align\\"><a href=\\"/contacts\\">contact us</a></li><li class=\\"left-align\\"><a href=\\"/about\\">about us</a></li></ul></header>"`;
